File: AdventOfCode2022/Python/Day07/Day07.py
import json

def get_inputs(file):
    """
    function to read in list of ints from a txt file
    """

    with open(file) as f:
        data = f.read().split('\n')

    return data

def Part1(filename):
    """
    Solutions to Part 1
    """
    data = get_inputs(filename)

    # format commands
    history = []
    current_command = []
    for line in data:
        # commands
        if line[0] == "$":
            # if current_command is empty
            if not current_command:
                current_command = [line, '']
            else:
                history.append(current_command)
                current_command = [line, '']
        # outputs 
        else:
            current_command[1] += line + ','
    history.append(current_command)

    # execute commands
    tree = {}
    cwd = ''

    for entry in history:
        command = entry [0]
        output = entry[1]

        # cd commands
        if command[:4] == "$ cd":

            cd_location = command[5:]

            # root
            if cd_location == "/":
                cwd = "/"
                tree["/"] = {}

            # move up a dir
            elif cd_location == "..":
                
                new_cwd = ''
                for dir in cwd.split('/')[:-2]:
                    new_cwd += dir +"/"
                cwd = new_cwd           

            # change to supplied location
            else:
                cwd += cd_location+"/"
            
        # list command
        elif command[:4] == "$ ls":

            for result in output.split(','):
                if result:
                    # dirs
                    if result[:3] == 'dir':
                        dir = result.split(' ')[1]
                        tree_str = '["/"]'
                        for dirp in (cwd.split('/')[1:-1]):
                            tree_str += f'["{dirp}"]'
                        eval(f'tree{tree_str}.update('+'{"'+dir+'":{}})')

                    # files
                    else:
                        filesize,file = result.split(' ')[0], result.split(' ')[1]
                        tree_str = '["/"]'
                        for dirp in (cwd.split('/')[1:-1]):
                            tree_str += f'["{dirp}"]'
                        eval(f'tree{tree_str}.update('+'{"'+file+'":'+filesize+'})')        
        
    print("cwd: ",cwd)
    print(json.dumps(tree, indent=4))
    print('#'*10)

    # count bytes
    results = []
    def function(dictt):
        keys = list(dictt.keys())
        vals = list(dictt.values())
        for key in keys:

            if type(dictt[key]) is dict:
                results.append({key: dictt[key]})
                function(dictt[key])

    function(tree)

    dirsizes = []
    for result in results:
        str_ = str(result)
        junk = list(str_.replace('{','').replace('}','').replace("'",'').replace(",",':').split(':'))
        total = 0
        for j in junk:
            try:
                total += int(j)
            except:
                continue
        
        print(list(result.keys())[0],total)
        dirsizes.append([list(result.keys())[0],total])
    
    # retrieve ans
    ans = 0 
    for dirsize in dirsizes:
        dir = dirsize[0]
        size = dirsize[1]

        if size <= 100000:
            ans += size
    
    print(f"Answer to Part 1 = {ans}")

    return dirsizes

def Part2(filename):
    """
    Solutions to Part 2
    """
    dirsizes = Part1(filename)

    used_size = dirsizes[0][1]
    total_size = 70000000
    required_size = 30000000

    options = []
    for dirsize in dirsizes:
        dir = dirsize[0]
        size = dirsize[1]

        if required_size <= (total_size - used_size + size):
            options.append(size)
    
    ans = min(options)
    print(f"Answer to Part 2 = {ans}")

def main():
    """
    main function
    """ 
    print("Advent-Of-Code 2022 - Day07")
    Part1("inputs.txt")
    Part2("inputs.txt")

if __name__ == "__main__":
    main()

File: AdventOfCode2022/Go/Day07/Day07.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type node struct {
	name     string
	size     int
	isFile   bool
	children map[string]*node
	parent   *node
}

func Parse(filename string) []map[string]string {

	// read file into string
	b, err := os.ReadFile(filename)
	if err != nil {
		fmt.Print(err)
	}

	lines := strings.Split(string(b), "\n")

	history := []map[string]string{}

	currentCommand := map[string]string{
		"command": "",
		"output":  "",
	}

	for _, line := range lines {
		if string(line[0]) == "$" {
			if currentCommand["command"] == "" {
				currentCommand["command"] = line
			} else {
				history = append(history, currentCommand)
				currentCommand = map[string]string{"command": line, "output": ""}
			}
		} else {
			currentCommand["output"] += (line + ",")
		}
	}
	history = append(history, currentCommand)

	return history
}

func calcSize(root node) (size int) {

	if root.isFile {
		return root.size

		// if dir sum file contents
	} else {
		for _, child := range root.children {
			size += calcSize(*child)
		}
	}
	return size
}

func Part1(filename string) int {

	history := Parse(filename)

	var cwd *node
	nodes := []*node{}

	for _, entry := range history {

		// cd commands
		if string(entry["command"][:4]) == "$ cd" {
			cdLocation := string(entry["command"][5:])

			// cd to root
			if cdLocation == "/" {
				cwd = &node{"/", 0, false, make(map[string]*node), nil}

				// cd to up a dir
			} else if cdLocation == ".." {
				cwd = cwd.parent

				// cd to a dir
			} else {
				cwd = cwd.children[cdLocation]
			}
		}

		// ls commands
		if string(entry["command"][:4]) == "$ ls" {

			for _, result := range strings.Split(entry["output"], ",") {

				// if dir contents not empty
				if result != "" {

					// nested dirs
					if string(result[:3]) == "dir" {
						dir := strings.Split(result, " ")[1]
						cwd.children[dir] = &node{dir, 0, false, make(map[string]*node), cwd}
						nodes = append(nodes, cwd.children[dir])

						// nested files
					} else {
						filesize, file := result[:strings.Index(result, " ")], result[strings.Index(result, " ")+1:]
						cwd.children[file] = &node{file, 0, true, make(map[string]*node), cwd}
						cwd.children[file].size, _ = strconv.Atoi(filesize)
					}
				}
			}
		}
	}

	var totalSize int

	for _, node := range nodes {

		size := calcSize(*node)

		// if size less than 100000 add to total
		if size <= 100000 {
			totalSize += size
		}
	}

	return totalSize
}

func Part2(filename string) int {

	history := Parse(filename)

	var cwd *node
	nodes := []*node{}

	for _, entry := range history {

		// cd commands
		if string(entry["command"][:4]) == "$ cd" {
			cdLocation := string(entry["command"][5:])

			// cd to root
			if cdLocation == "/" {
				cwd = &node{"/", 0, false, make(map[string]*node), nil}
				nodes = append(nodes, cwd)

				// cd to up a dir
			} else if cdLocation == ".." {
				cwd = cwd.parent

				// cd to a dir
			} else {
				cwd = cwd.children[cdLocation]
			}
		}

		// ls commands
		if string(entry["command"][:4]) == "$ ls" {

			for _, result := range strings.Split(entry["output"], ",") {

				// if dir contents not empty
				if result != "" {

					// nested dirs
					if string(result[:3]) == "dir" {
						dir := strings.Split(result, " ")[1]
						cwd.children[dir] = &node{dir, 0, false, make(map[string]*node), cwd}
						nodes = append(nodes, cwd.children[dir])

						// nested files
					} else {
						filesize, file := result[:strings.Index(result, " ")], result[strings.Index(result, " ")+1:]
						cwd.children[file] = &node{file, 0, true, make(map[string]*node), cwd}
						cwd.children[file].size, _ = strconv.Atoi(filesize)
					}
				}
			}
		}
	}

	var totalSize int

	for _, node := range nodes {

		if node.name == "/" {
			totalSize = calcSize(*node)
		}
	}

	// dumb var for comparison
	answer := 100000000000000000

	for _, node := range nodes {

		size := calcSize(*node)

		if size < answer && 70000000-(totalSize-size) >= 30000000 {
			answer = size
		}
	}

	return answer
}

func main() {

	testfile := "tests.txt"
	inputfile := "inputs.txt"

	fmt.Println("Advent-Of-Code 2022 - Day07")
	fmt.Printf("Tests : Answer to Part 1 = %v\n", Part1(testfile))
	fmt.Printf("Inputs: Answer to Part 1 = %v\n", Part1(inputfile))
	fmt.Printf("Tests : Answer to Part 2 = %v\n", Part2(testfile))
	fmt.Printf("Inputs: Answer to Part 2 = %v\n", Part2(inputfile))
}
